Firestore/core/src/firebase/firestore/remote/stream_operation.h

```diff
@@ -0,0 +1,184 @@
+#include "grpcpp/generic/generic_stub.h"
+#include "grpcpp/support/byte_buffer.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+class GrpcStream;
+
+/**
+ * An operation that notifies the corresponding `GrpcStream` on its completion.
+ *
+ * All created operations are always put on the GRPC completion queue (see
+ * `ExecuteOperation` factory function). Operation expects that once it's
+ * received back from the GRPC completion queue, `Complete()` will be called on
+ * it. `Complete` doesn't notify the observing stream immediately; instead, it
+ * schedules the notification on the Firestore async queue. If the stream
+ * doesn't want to be notified, it should call `UnsetObserver` on the operation.
+ *
+ * Operation is "self-owned"; operation deletes itself in its `Complete` method.
+ *
+ * Operation expects all GRPC objects pertaining to the current stream to remain
+ * valid until the operation comes back from the GRPC completion queue.
+ */
+class StreamOperation : public GrpcOperation {
+ public:
+  // TODO(varconst): make call and queue getters on stream to minimize passing
+  // data around?
+  StreamOperation(GrpcStream* observer,
+                  grpc::GenericClientAsyncReaderWriter* call,
+                  util::AsyncQueue* firestore_queue);
+
+  // This factory function, together with the fact that constructors of all
+  // derived operations are private, ensures that each created operation is
+  // always put on the GRPC completion queue.
+  template <typename Op, typename... Args>
+  static Op* ExecuteOperation(GrpcStream* observer,
+                              grpc::GenericClientAsyncReaderWriter* call,
+                              util::AsyncQueue* firestore_queue,
+                              Args... args) {
+    Op* op = new Op{observer, call, firestore_queue, std::move(args)...};
+    op->Execute();
+    return op;
+  }
+
+  /**
+   * Puts the operation on the GRPC completion queue.
+   *
+   * Must be called on the Firestore async queue.
+   */
+  void Execute() override;
+
+  /**
+   * Marks the operation as having come back from the GRPC completion queue and
+   * puts notifying the observing stream on the Firestore async queue. The given
+   * `ok` value indicates whether the operation completed successfully.
+   *
+   * This function deletes the operation.
+   *
+   * Must be called outside of Firestore async queue.
+   */
+  void Complete(bool ok) override;
+
+  void UnsetObserver();
+
+  // This is a blocking function; it blocks until the operation comes back from
+  // the GRPC completion queue. It is important to only call this function when
+  // the operation is sure to come back from the queue quickly.
+  void WaitUntilOffQueue();
+  std::future_status WaitUntilOffQueue(std::chrono::milliseconds timeout);
+
+ private:
+  virtual void DoExecute(grpc::GenericClientAsyncReaderWriter* call) = 0;
+  virtual void DoComplete(GrpcStream* observer) = 0;
+
+  GrpcStream* observer_ = nullptr;
```

Comment:
This name is confusing since there's already `GrpcStreamObserver` type and this isn't one of them. This would be better named `stream_`.

Reply:
Done.